### DISCUSSION TITLE
TypeError: router.events is undefined in NextJS 13

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#50630] [App Router]
I'm trying to show Loading text while the route is loading the url state using the latest **nextjs 13**, **next/router** replaced with **next/navigation**

``` js
"use client";
import React, { useState, useEffect } from "react";
import { useRouter } from "next/navigation";
import Link from "next/link";
import axios from "axios";

const Header = () => {
  const [searchQuery, setSearchQuery] = useState("");
  const [isLoading, setIsLoading] = useState(false); 
  const router = useRouter();

  useEffect(() => {
    const handleStart = () => {
      setIsLoading(true); 
    };

    const handleComplete = () => {
      setIsLoading(false); 
    };

    router.events.on("routeChangeStart", handleStart);
    router.events.on("routeChangeComplete", handleComplete);

    return () => {
      router.events.off("routeChangeStart", handleStart);
      router.events.off("routeChangeComplete", handleComplete);
    };
  }, [router]);

  const handleFormSubmit = async (e) => {
    e.preventDefault();
    setIsLoading(true); 

    try {
      const response = await axios.get(`/result?search_query=${encodeURIComponent(searchQuery)}`);
      const data = response.data;


      console.log("Search Results:", data);

      router.push(`/result?search_query=${encodeURIComponent(searchQuery)}`);
    } catch (error) {
      console.log("Error:", error);
    } finally {
      setIsLoading(false); 
    }
  };

  const handleInputChange = (e) => {
    setSearchQuery(e.target.value);
  };

  return (
    {/* Render data */}
  );
};

export default Header;
```

Using **next/router** gives **Error: NextRouter was not mounted**.

My file is in **app/components/Header.jsx**, changing to client side to **pages/Header.jsx** also not working

**Comment by icyJoseph** [comment]
There's no router events in App Dir's router. At least not for now: https://github.com/vercel/next.js/discussions/42016

**Reply by ghost** [reply]
Is there any relevant method to achieve things like router.events? Or maybe downgrading to Next 11 or below 13 is the final solution?

**Reply by icyJoseph** [reply]
I've been trying to hack something together, but the best I've seen are listed here:

- https://github.com/vercel/next.js/discussions/42016
- https://github.com/vercel/next.js/discussions/41745#discussioncomment-3987025

I had hoped route interceptors could provide something similar, but I failed to get the loading to work. Perhaps it'll be possible soon, but AFAIK, not yet.
